TauAnalysis/MCEmbeddingTools/plugins/EmbeddingLHEProducer.cc

```diff
@@ -435,41 +410,27 @@ void EmbeddingLHEProducer::InitialRecoCorrection(TLorentzVector &positiveLepton,
     return;
   edm::LogInfo("TauEmbedding") << "Applying initial reconstruction correction";
   TLorentzVector Z = positiveLepton + negativeLepton;
-
   edm::LogInfo("TauEmbedding") << "MuMinus before. Pt: " << negativeLepton.Pt() << " Mass: " << negativeLepton.M();
-  // std::cout << " MuMinus before. Pt: " << negativeLepton.Pt() << " Mass: " << negativeLepton.M() << " Energy: " << negativeLepton.E() << std::endl;
   float diLeptonMass = (positiveLepton + negativeLepton).M();
   if (diLeptonMass > 60. && diLeptonMass < 122.) {
-    // std::cout << "DiLeptonMass " << diLeptonMass << std::endl;
-    float zmass = 91.1876;
+    static constexpr float zmass = 91.1876;
     float correction_deviation =
         5.;  // to ensure only a correction that drops corresponding to a Gaussian with mean zmass and std. dev. 5 GeV
     double EmbeddingCorrection =
         1.138;  // value derived by function fitting to fold embedded mass spectrum back to original selection when using mu -> mu embedding
-    EmbeddingCorrection =
-        EmbeddingCorrection /
+    EmbeddingCorrection /=
         (EmbeddingCorrection -
          (EmbeddingCorrection - 1.) * exp(-pow((diLeptonMass - zmass), 2.) / (2. * pow(correction_deviation, 2.))));
     EmbeddingCorrection = ((diLeptonMass + (EmbeddingCorrection - 1.) * zmass) / (diLeptonMass * EmbeddingCorrection));
     double correctedpositiveLeptonEnergy = std::sqrt(
-        muonMass_ * muonMass_ + EmbeddingCorrection * positiveLepton.Px() * EmbeddingCorrection * positiveLepton.Px() +
-        EmbeddingCorrection * positiveLepton.Py() * EmbeddingCorrection * positiveLepton.Py() +
-        EmbeddingCorrection * positiveLepton.Pz() * EmbeddingCorrection * positiveLepton.Pz());
+        (pow(muonMass_, 2) / pow(EmbeddingCorrection, 2))  + pow(positiveLepton.Px(), 2) + pow(positiveLepton.Py(), 2) + pow(positiveLepton.Pz(), 2));
```

Comment:
Maybe this is leading to the changes we observed?

While the new calculation seems to do things better, it might well be, that due to numerics, this leads to different results?

Reply:
I checked this, is results in identical 4 vectors, comparing this and the previous implementation. No numerical differences

Reply:
OK thanks for clarification!


I don't understand then, what @winterchristian has shown today in his presentation. Since all other changes look fine

Reply:
In fact, the changes in the tau p_T that I showed in the presentation are due to this change. I checked, and the minor precision changes introduced here lead to large changes in the p_T distribution after simulation. But that's fine, since there is no right or wrong with numerical precision uncertainties and besides, this is a random process.